### ISSUE TITLE
[AMD][kernel][Tier 6] FlyDSL-generated FA-2 backward pass for training on MI355X

### ISSUE THREAD

**Issue by sunxxuns** [amdpilot-org/FlyDSL#31]
## Goal

Use **FlyDSL to generate the Flash-Attention backward pass** for
training on MI355X. Today FlyDSL ships `kernels/flash_attn_func.py`
(forward) but no backward — every training stack (LlamaFactory,
OpenRLHF, verl, Megatron-LM-AMD) that uses FA forward on AMD has to
fall back to PyTorch autograd for the backward, which is 5-10× slower
than a fused FA backward and forces re-materialisation of softmax
statistics.

The companion issue for the GFX9-specific MFMA mainloop is
amdpilot-org/tilelang#2 — this FlyDSL issue is for the *DSL-generated*
variant that can be re-targeted to different head_dims via the Fly
dialect's layout algebra. The two should produce numerically-equivalent
gradients on the same inputs.

Reuses cached **`amd/Llama-3.3-70B-Instruct-MXFP4-Preview`** and
**`Qwen/Qwen3-32B`** for end-to-end gradient sanity check at
training-realistic shapes.

## Hardware / workload

```
Hardware:    1 × MI355X (per-kernel parity), 8 × MI355X (end-to-end)
Shape:       (B=2, H=16, S=4096, D=128) bf16 Q/K/V + bf16 dOut
             — Llama-3.3-70B GQA-grouped shape
Baseline:    PyTorch autograd backward through
             `torch.nn.functional.scaled_dot_product_attention`
Target:      kernel ms < 0.4 × autograd baseline at
             max_abs_diff(dQ,dK,dV) < 1e-3 (fp32 accumulation),
             < 5e-3 (bf16 accumulation).
```

## Why FlyDSL on MI355X

FA backward's recompute schedule needs three GEMM stages per output
tile (dS, attention recompute, dQ/dK/dV accumulate). The Fly dialect's
layout algebra (composition / partition / coord mapping) expresses
the recompute tile reuse cleanly — Triton needs hand-tuned LDS pad
shapes for each (B, H, D) combination, while FlyDSL can produce
a single kernel that auto-tunes the partition.

## Scope

1. Add `kernels/flash_attn_bwd.py` implementing the FA-2 backward
   schedule with the Fly partition primitives. Single head_dim=128 +
   bf16, no-causal first; causal-mask + head_dim ∈ {64, 256} follow.
2. Wire dispatch in `python/flydsl/__init__.py` so the autograd hook
   `torch.autograd.Function.backward` on `flash_attn_func` calls into
   the new kernel.
3. Parity test in `tests/kernels/test_flash_attn_bwd.py`: tiny shape
   (B=1 H=2 S=128 D=64) vs PyTorch autograd reference;
   max_abs_diff < 1e-3 per dQ/dK/dV.
4. End-to-end gradient sanity in
   `tests/integration/test_llama33_70b_bwd_step.py`: 1-step backward
   through a 2-layer attention stack on Llama-3.3-70B (single batch,
   S=512), assert finite gradients with no NaN/Inf.
5. Benchmark `tests/kernels/bench_flash_attn_bwd.py` reporting
   `bwd_ms_per_iter` for the spec shape.

## Acceptance

- Parity test passes.
- Benchmark on MI355X reports `bwd_ms_per_iter < 0.4 × autograd_ref`.
- E2E sanity passes (finite grads on 2-layer Llama-3.3 stack).
- Gradient differs from amdpilot-org/tilelang#2 reference by <
  1e-3 (the two implementations should converge on the same answer).

**Comment by jhinpan** [comment]
**AMDPilot — accumulated progress on this issue**

_Auto-generated from successive trials. Latest update: 2026-05-23T17:18:44Z by `issue-FlyDSL-31-r23`._

**Best result so far**: 2.0600 ms (-19.5% vs baseline 2.5567)
**Achieved by**: ``issue-FlyDSL-31-r17``
**Key changes that mattered**: Flash-Attn2 backward kernel achieves 1.16x speedup over SDPA

### Verified infra fixes (apply during Stage 0)

- Resolved flydsl.expr import blocker enabling test_harness.py execution
- Set PYTHONPATH=/workspace/FlyDSL/python:/workspace/FlyDSL to resolve flydsl.expr ModuleNotFoundError


### Working approaches and their best metric

- Identified launch_flash_attn_func and _auto_launch as critical entry points via static analysis
- Achieved first successful small FlyDSL forward kernel launch (Max output: 0.6640625)
- Resolved pybind11_type TypeError to enable SDPBackend enumeration
- Executed full backward benchmark confirming Flash-Attn2 superiority (2.06ms vs 2.39ms)


### Approaches that did NOT help — don't retry these

- Retrying agent with identical 'no source edits' failure without changing prompt or environment (Trials 3-4)
- Setting HSA_OVERRIDE_GFX_VERSION=9.4.0 — caused 'No HIP GPUs available' error
- Blindly tweaking env vars (HSA_ENABLE_SDMA, GPU_MAX_HW_QUEUES, AMD_LOG_LEVEL) — failed to resolve segfault
- Re-verifying hipcc/ROCm versions after initial confirmation — redundant and unproductive
- Setting HSA_OVERRIDE_GFX_VERSION=9.4.0 — triggered 'No HIP GPUs available' error
- Tweaking HSA_ENABLE_SDMA, GPU_MAX_HW_QUEUES, AMD_LOG_LEVEL — failed to resolve segfault
- Hallucinating base image 'amdpilot/issue-flydsl-29-r3:stage0' caused immediate bundle validation failure
- Re-verifying torch.randn and rewriting profile_bwd.py instead of parsing the existing trace JSON generated during the crash
- Switching HIP_VISIBLE_DEVICES to index 2 — no change, crash persists on node smci355-ccs-aus-n08-09
- Gathering static env dumps (rocm-smi, ldconfig) instead of kernel traces — failed to resolve segfault
- Retrying CUDA allocation after confirming low-power state warning — hardware remains unresponsive
- Using grep with shell globs on flydsl/expr/*.py failed due to expansion issues, slowing file discovery


### Open ideas to try next

- Update gpu-profiling skill to instruct agents to check for existing *.json trace artifacts immediately after script failure
- Parse /workspace/traces/profile_bwd_trace.json to analyze the latency gap between baseline and optimized kernels
- Create minimal reproduction script using identified launcher functions
- Analyze existing IR dumps in /root/.flydsl/debug/ for regression patterns
- Proceed to backward pass implementation now that forward context is stable
- Enhance ExecutionMonitor to detect 'low-power state' warnings and trigger node-level GPU reset
- Request container rebuild or node re-lease due to persistent wedged GPU context
- Force GPU wake-up sequence before allowing further CUDA allocation attempts
- Populate flydsl.kernels package to expose public symbols
- Generate mandatory reproduce.json and reproduce.md artifacts
- Inject pre-computed symbol index into agent context to replace manual grep/sed exploration loops
- Proceed to backward pass implementation now that forward context and imports are stable


_amdpilot bot. Edit between the markers to add/correct fixes; the next update will preserve your changes. To disable updates on this issue, replace the body with the marker line only._

**Comment by sunxxuns** [comment]
## Closing as not tractable by the current autonomous agent (2026-05-23)

This issue has accumulated 6+ chronic-fail retries with 0 wins over the past 12 h on
the autonomous loop. Triage: FA-2 backward via FlyDSL Tier 6 — agent cannot produce correct recompute schedule.

The agent (Kimi-K2.6) does not have sufficient AMD gfx950/gfx942 MFMA + LDS-bank-conflict
expertise to make iterative progress on this scope autonomously. Every retry is
landing on the same blocker (lack of arch-specific reasoning), wasting fleet GPU time.

**To re-open**:
- Pre-decompose into one of: "implement a tile-shape sweep harness" (which the agent CAN do),
  or "verify numerical parity with an upstream reference at single-shape S" (small + provable).
- OR escalate to a human kernel engineer; the broad multi-kernel scope here is not autonomous-friendly.

Closing this round so the fleet doesn't keep burning retry cycles. The work itself
is still valuable — the wrapping just needs to be different.
